### PR TITLE
Add compatibility to vcpkg buildsystem

### LINKED ISSUE #7642: Introduce vcpkg for easier development on Windows
##### Issue type
 
- Feature request

##### Minetest version
5+

##### OS / Hardware
Operating system: Windows

##### Summary
Hi folks, hows everyone been?
I just learned about Vcpkg. I think this could help us a lot in making the windows binaries.

> **You'll have to forgive the crudeness of this model, I didn't have time to paint it or build it to scale.**

> Quick Start
> Vcpkg helps you get C and C++ libraries on Windows. This tool and ecosystem are currently in a preview state; your involvement is vital to its success.

The following was take [from here](https://vcpkg.readthedocs.io/en/latest/examples/using-sqlite/).

#### Step 1: Install
First, we need to know what name [Sqlite](https://sqlite.org/) goes by in the ports tree. To do that, we'll run the `search` command and inspect the output:
```
S D:\src\vcpkg> .\vcpkg search sqlite
libodb-sqlite        2.4.0            Sqlite support for the ODB ORM library
sqlite3              3.15.0           SQLite is a software library that implements a se...

If your library is not listed, please open an issue at:
    https://github.com/Microsoft/vcpkg/issues
```
Looking at the list, we can see that the port is named "sqlite3". You can also run the `search` command without arguments to see the full list of packages.

Installing is then as simple as using the `install` command.
```
PS D:\src\vcpkg> .\vcpkg install sqlite3
-- CURRENT_INSTALLED_DIR=D:/src/vcpkg/installed/x86-windows
-- DOWNLOADS=D:/src/vcpkg/downloads
-- CURRENT_PACKAGES_DIR=D:/src/vcpkg/packages/sqlite3_x86-windows
-- CURRENT_BUILDTREES_DIR=D:/src/vcpkg/buildtrees/sqlite3
-- CURRENT_PORT_DIR=D:/src/vcpkg/ports/sqlite3/.
-- Downloading https://sqlite.org/2016/sqlite-amalgamation-3150000.zip...
-- Downloading https://sqlite.org/2016/sqlite-amalgamation-3150000.zip... OK
-- Testing integrity of downloaded file...
-- Testing integrity of downloaded file... OK
-- Extracting source D:/src/vcpkg/downloads/sqlite-amalgamation-3150000.zip
-- Extracting done
-- Configuring x86-windows-rel
-- Configuring x86-windows-rel done
-- Configuring x86-windows-dbg
-- Configuring x86-windows-dbg done
-- Build x86-windows-rel
-- Build x86-windows-rel done
-- Build x86-windows-dbg
-- Build x86-windows-dbg done
-- Package x86-windows-rel
-- Package x86-windows-rel done
-- Package x86-windows-dbg
-- Package x86-windows-dbg done
-- Performing post-build validation
-- Performing post-build validation done
Package sqlite3:x86-windows is installed
```
We can check that sqlite3 was successfully installed for x86 windows desktop by running the `list` command.
```
PS D:\src\vcpkg> .\vcpkg list
sqlite3:x86-windows         3.15.0           SQLite is a software library that implements a se...
```
To install for other architectures and platforms such as Universal Windows Platform or x64 Desktop, you can suffix the package name with :`<target>`.
`PS D:\src\vcpkg> .\vcpkg install sqlite3:x86-uwp zlib:x64-windows`
See `.\vcpkg help triplet` for all supported targets.
##### VS/MSBuild Project (User-wide integration)
The recommended and most productive way to use vcpkg is via user-wide integration, making the system available for all projects you build. The user-wide integration will prompt for administrator access the first time it is used on a given machine, but afterwords is no longer required and the integration is configured on a per-user basis.
```
PS D:\src\vcpkg> .\vcpkg integrate install
Applied user-wide integration for this vcpkg root.

All C++ projects can now #include any installed libraries.
Linking will be handled automatically.
Installing new libraries will make them instantly available.
```
_Note: You will need to restart Visual Studio or perform a Build to update intellisense with the changes._
You can now simply use File -> New Project in Visual Studio 2015 or Visual Studio 2017 and the library will be automatically available. For Sqlite, you can try out their [C/C++ sample](https://sqlite.org/quickstart.html).

To remove the integration for your user, you can use `.\vcpkg integrate remove`.
##### CMake (Toolchain File)
The best way to use installed libraries with cmake is via the toolchain file  `scripts\buildsystems\vcpkg.cmake`. To use this file, you simply need to add it onto your CMake command line as `-DCMAKE_TOOLCHAIN_FILE=D:\src\vcpkg\scripts\buildsystems\vcpkg.cmake`.

If you are using CMake through Open Folder with Visual Studio 2017 you can define  `CMAKE_TOOLCHAIN_FILE` by adding a "variables" section to each of your `CMakeSettings.json` configurations:
```
{
  "configurations": [{
    "name": "x86-Debug",
    "generator": "Visual Studio 15 2017",
    "configurationType" : "Debug",
    "buildRoot":  "${env.LOCALAPPDATA}\\CMakeBuild\\${workspaceHash}\\build\\${name}",
    "cmakeCommandArgs": "",
    "buildCommandArgs": "-m -v:minimal",
    "variables": [{
      "name": "CMAKE_TOOLCHAIN_FILE",
      "value": "D:\\src\\vcpkg\\scripts\\buildsystems\\vcpkg.cmake"
    }]
  }]
}
```
Now let's make a simple CMake project with a main file.
```
# CMakeLists.txt
cmake_minimum_required(VERSION 3.0)
project(test)

find_package(Sqlite3 REQUIRED)

add_executable(main main.cpp)
target_link_libraries(main sqlite3)
```
```
// main.cpp
#include <sqlite3.h>
#include <stdio.h>

int main()
{
    printf("%s\n", sqlite3_libversion());
    return 0;
}
```
Then, we build our project in the normal CMake way:
```
PS D:\src\cmake-test> mkdir build 
PS D:\src\cmake-test> cd build
PS D:\src\cmake-test\build> cmake .. "-DCMAKE_TOOLCHAIN_FILE=D:\src\vcpkg\scripts\buildsystems\vcpkg.cmake"
    // omitted CMake output here //
-- Build files have been written to: D:/src/cmake-test/build
PS D:\src\cmake-test\build> cmake --build .
    // omitted MSBuild output here //
Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:02.38
PS D:\src\cmake-test\build> .\Debug\main.exe
3.15.0
```
_Note: The correct sqlite3.dll is automatically copied to the output folder when building for x86-windows. You will need to distribute this along with your application._
###### Handling libraries without native cmake support
Unlike other platforms, we do not automatically add the `include\` directory to your compilation line by default. If you're using a library that does not provide CMake integration, you will need to explicitly search for the files and add them yourself using `find_path()` and `find_library()`.
```
# To find and use catch
find_path(CATCH_INCLUDE_DIR catch.hpp)
include_directories(${CATCH_INCLUDE_DIR})

# To find and use azure-storage-cpp
find_path(WASTORAGE_INCLUDE_DIR was/blob.h)
find_library(WASTORAGE_LIBRARY wastorage)
include_directories(${WASTORAGE_INCLUDE_DIR})
link_libraries(${WASTORAGE_LIBRARY})

# Note that we recommend using the target-specific directives for a cleaner cmake:
#     target_include_directories(main ${LIBRARY})
#     target_link_libraries(main PRIVATE ${LIBRARY})
```

### PR DESCRIPTION
**Pull request by adrido**
Vcpkg is a C++ Library Manager for Windows, Linux, and MacOS. https://github.com/Microsoft/vcpkg

It is really helpful because it makes it easy to manage Minetest's load of (optional) dependencies. Especially on Windows.

This PR changes all required things to make cmake+vcpkg work out of the box.
It also updates the documentation and adds instructions how to build on windows.

Here a checklist what's currently possible:

Triplets:
✔ x64-windows
❌ x64-windows-static (link error with irrlicht, openAL, curl and leveldb)
✔ x86-windows
❓ x86-windows-static (needs testing)

Required Features:
✔ build client
✔ build server
✔ irrlicht
✔ sqlite3
✔ zlib

Optional Features:
✔ curl
❓  curses (have no idea what's that)
✔ freetype
❓  gettext (supported by vcpkg but not tested)
❓  gles (probably not possible)
✔ leveldb
✔ luajit
✔ postgresql (changes to Minetest required)
❌ redis (Minetest depends on 'hiredis' which is Unix exclusive)
✔ sound
❓ spatial (there is a libary called libspatiallite)
❓ system-gmp (not supported by vcpkg)
❓ system-jsoncpp (supported by vcpkg but not tested)

It would be nice if someone could follow the documentation and find out if its complete and understandable.

That missing optional features can also added later, so I would consider this PR as ready for review.
This PR already includes #8295 
Fixes #7642 

Edit 14.03.2019: replaced the task list with "emotes" as it was not really a task list

### CODE REVIEW

**Comment by SmallJoker** [comment]
> Should the old outdated windows build instructions kept or removed?

I think you can remove those. If someone still wants to compile the annoying way, then there's still the same guide on the wiki (maybe add a link to it?).

> or schould it stay like it is (lua51.lib) this would make it impossible use it under Linux and OS-X

Guard it with `if`s, to keep compatibility as much as possible.

**Comment by adrido** [comment]
Done. Regarding PostgreSQL:
https://github.com/minetest/minetest/blob/77961aa73d1947a2eabaa832319b7d73db6d13c2/src/CMakeLists.txt#L161-L186
BTW is it intentional that 
first `find_library(POSTGRESQL_LIBRARY pq)` is called 
and then `find_package("PostgreSQL")` but at the end linking to POSTGRESQL_LIBRARY??? https://github.com/minetest/minetest/blob/master/src/CMakeLists.txt#L596

in case of vcpkg it looks like this:
`find_library(POSTGRESQL_LIBRARY pq)` --> POSTGRESQL_LIBRARY-NOTFOUND (its named libpq)
`find_package("PostgreSQL")` --> found --> sets PostgreSQL_LIBRARY to libpq.lib
`target_link_libraries(${PROJECT_NAME} ${POSTGRESQL_LIBRARY})` --> ERROR

**Comment by adrido** [comment]
Rebased.

**Comment by adrido** [comment]
Rebased. As https://github.com/minetest/minetest/pull/8435 is merged, PostgreSQL is now working as well.

File: README.md [hunk 1]
```diff
@@ -284,165 +284,77 @@ Library specific options:
...
+* Visual Studio 2015 or newer https://visualstudio.microsoft.com
+* CMake https://cmake.org/download/
+* vcpkg https://github.com/Microsoft/vcpkg
+* git https://git-scm.com/downloads
+
+### Compiling and Installing the dependencies
+
+Minetest relays on varios libaries. some are optional and only provide different database backends, and some are mandatory.
+
+It is highly recommended to use vcpkg as package manager. Of course you can compile the dependencies on your own, but you have to use different buildsystems, solve the dependencies of the dependencies, and may just end up with some linker errors.
```
Comment (ClobberXD) [inline README.md]: The second sentence is too long and could be omitted IMHO - it just adds to the previous sentence, and doesn't add too much meaning. The README has to be clear and concise. :)

File: README.md [hunk 2]
```diff
@@ -284,165 +284,77 @@ Library specific options:
...
+### Requirements
+
+* Visual Studio 2015 or newer https://visualstudio.microsoft.com
+* CMake https://cmake.org/download/
+* vcpkg https://github.com/Microsoft/vcpkg
+* git https://git-scm.com/downloads
+
+### Compiling and Installing the dependencies
+
+Minetest relays on varios libaries. some are optional and only provide different database backends, and some are mandatory.
```
Comment (ClobberXD) [inline README.md]: This sentence is _somewhat_ redundant, as the `#### a) Using vcpkg to install dependencies` section already lists the optional ones.

File: README.md [hunk 3]
```diff
@@ -284,165 +284,77 @@ Library specific options:
...
+This is outdated and not recommended. Follow the instructions on https://dev.minetest.net/Build_Win32_Minetest_including_all_required_libraries#VS2012_Build
+
+### Compile Minetest
+
+#### a) Using the vcpkg toolchain and CMakeGUI
+1. Start up the CMake GUI
+2. Select "Browse Source..." and select DIR/minetest
+3. Select "Browse Build..." and select DIR/minetest-build
+4. Select "Configure"
+5. Choose the right visual Studio version and target platform. It have to fit to the target of the installed dependencies
```
Comment (ClobberXD) [inline README.md]:
```suggestion
5. Choose the right visual Studio version and target platform. It has to match the version of the installed dependencies
```

File: README.md [hunk 4]
```diff
@@ -284,165 +284,77 @@ Library specific options:
...
+9. Click Finish
+10. Wait until cmake have generated the cash file
+11. If there are any errors, solve them and hit "Configure"
+12. Click "Generate"
+13. Click "Open Project"
+14. Compile Minetest inside Visual studio.
+
+#### b) Using the vcpkg toolchain and the commandline
+
+Its probably the best, to just create a bat or ps1 file. But you can just paste this two lines inside the powershell window too.
```
Comment (ClobberXD) [inline README.md]:
```suggestion
Run the following script in Powershell:
```

File: README.md [hunk 5]
```diff
@@ -284,165 +284,77 @@ Library specific options:
...
+
+```powershell
+cmake . -G"Visual Studio 15 2017 Win64" -DCMAKE_TOOLCHAIN_FILE=D:/vcpkg/scripts/buildsystems/vcpkg.cmake -DCMAKE_BUILD_TYPE=Release -DENABLE_GETTEXT=0 -DENABLE_CURSES=0
+cmake --build . --config Release
+```
+Make shure that the right Compiler is selected and the path to the vcpkg toolchain file is correct.
```
Comment (ClobberXD) [inline README.md]:
```suggestion
Make sure that the right compiler is selected and the path to the vcpkg toolchain is correct.
```

File: README.md [hunk 6]
```diff
@@ -284,165 +284,77 @@ Library specific options:
...
+
+#### c) Using your own compiled libaries
+
+This is outdated and not recommended. Follow the instructions on https://dev.minetest.net/Build_Win32_Minetest_including_all_required_libraries#VS2012_Build
```
Comment (ClobberXD) [inline README.md]:
```suggestion
**This is outdated and not recommended**

Follow the instructions on https://dev.minetest.net/Build_Win32_Minetest_including_all_required_libraries#VS2012_Build
```

File: README.md (hunk 1 above)
Comment (adrido) [reply README.md]:
```suggestion
You can compile the libraries on your own, but it is highly recommended to use vcpkg as package manager.
```

File: README.md (hunk 2 above)
Comment (adrido) [reply README.md]:
```suggestion
Minetest relays on various third party libraries. 
```
Thanks for your review. 
What do you think about my suggestions?
I does not want to omit the paragraph completely as there is missing a introduction then.

Comment (ClobberXD) [reply README.md]: But is there a need for this section at all? It's quite obvious MT depends on various third-party libs. :)

File: README.md (hunk 1 above)
Comment (ClobberXD) [reply README.md]: Documentation uses passive voice, and pronouns like "you" and "we" wouldn't sound right. Is this sentence really needed? Again, the README is the official front page of a project, and overly wrong sentences won't make a good impression on the readers.

File: README.md (hunk 2 above)
Comment (adrido) [reply README.md]: You are right.

File: src/CMakeLists.txt [hunk 7]
```diff
@@ -251,32 +251,23 @@ if(WIN32)
...
+	# Dll's are automatically copied to the output directory by vcpkg when VCPKG_APPLOCAL_DEPS=ON
+	if(NOT VCPKG_APPLOCAL_DEPS)
+		find_file(ZLIB_DLL NAMES "zlib.dll" "zlib1.dll" DOC "Path to zlib.dll for installation (optional)")
+		if(ENABLE_SOUND)
+			set(OPENAL_DLL "" CACHE FILEPATH "Path to OpenAL32.dll for installation (optional)")
+			set(OGG_DLL "" CACHE FILEPATH "Path to libogg.dll for installation (optional)")
+			set(VORBIS_DLL "" CACHE FILEPATH "Path to libvorbis.dll for installation (optional)")
+			set(VORBISFILE_DLL "" CACHE FILEPATH "Path to libvorbisfile.dll for installation (optional)")
+		endif()
+		if(USE_LUAJIT)
```
Comment (SmallJoker) [inline src/CMakeLists.txt]: What happened to freetype?

Comment (adrido) [reply src/CMakeLists.txt]: Already handled there
https://github.com/minetest/minetest/blob/master/src/CMakeLists.txt#L137

**Comment by rubenwardy** [comment]
This needs testing to make sure the buildbot script still works, and other ways of compiling on Windows too

**Pull request merged**